### PR TITLE
Performance Tuning

### PR DESCRIPTION
**Pull request by l-monninger**
# Summary
- **RFCs**: $\emptyset$
- **Categories**: `protocol-units`

General branch for performance tuning experimentation.

# Changelog

# Testing

# Outstanding issues

### CODE REVIEW

File: util/movement-types/src/lib.rs
Comment (mzabaluev) [inline util/movement-types/src/lib.rs:150]: Good; I thought to ask about the choice of the hasher for this and other IDs.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (musitdev) [inline protocol-units/da/m1/light-node/src/v1/sequencer.rs:266]: This is a good example of how this way to define everything async can lead to inefficient code. This code block should be spawn_blocking because there's no async IO in the memseq publish process.
The` memseq.publish()` call `add_transaction` which came from the async trait `MempoolTransactionOperations`. And in the end it calls `RocksdbMempool::add_mempool_transaction()` which you `spawn_blocking` like for DB.
So instead of spawning one thread for all transaction to publish we spawn one thread for each transaction which can lead to worse performance than spawning nothing and with the current code it's very hard to change.
@mzabaluev, the new architecture should take care of that and async should be dissociated from sync function.
And I think it's for the same reason that you `spawn_blocking` each method of DB. I think they shouldn't be `spawn_blocking` because each method do a simple thing, and it's more the caller that could decide to `spawn_blocking` or not.
`spawn_blocking` too much short task can be less efficient than not doing it.

**Review by musitdev — COMMENTED** [review]
Will see if it's better for the performance, but I think we face the limit of the optimization of the current code without doing important refactoring.
Before merging, the performance should be tested because not sure if it helps for it if the number of Tx increase.

File: .github/workflows/checks.yml
Comment (l-monninger) [reply .github/workflows/checks.yml:56]: Yeah, because of the Mocha sync times we were just running against Holesky. But, this is dropping this check for the moment.

File: docker/compose/suzuka-full-node/docker-compose.yml
Comment (l-monninger) [reply docker/compose/suzuka-full-node/docker-compose.yml:39]: There hasn't been. But, a restart should be reasonable.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs:71]: Yeah, but the trouble here is that I don't known the `block_id` and that's what I actually wanted to start tracking. 

I'm sure there's some way to do this with like a closure for the `span`, or by getting the `id` of the span then logging something with the `block_id` thereafter. Would appreciate pointers here.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs:74]: So, does the `parent:` allow me to add information like the `block_id`?

I wonder how this nesting will be to deal with in the Jupyter notebook. Shouldn't be too bad I supposed, just one `group_by` away.

File: protocol-units/execution/opt-executor/src/executor/execution.rs
Comment (l-monninger) [reply protocol-units/execution/opt-executor/src/executor/execution.rs:113]: So... if we want to get rid of the mempool lock, this basically needs to happen. 

In a previous usage of the mempool--like way back a couple months ago--this resolved an issue with transaction confirmations. Now, however, it does not seem to be necessary. 

I think this is mainly because we've stopped using the mempool for anything other than checking that a transaction passes the basic mempool validity checks. All of the actual transaction batching and whatnot is just passing through pipe.

File: protocol-units/execution/opt-executor/src/executor/initialization.rs
Comment (l-monninger) [reply protocol-units/execution/opt-executor/src/executor/initialization.rs:131]: See the comment above. Everything is fairly temporary for the moment. But, it doesn't seem we need to hold the mempool reference any longer.

File: protocol-units/execution/opt-executor/src/executor/initialization.rs
Comment (l-monninger) [reply protocol-units/execution/opt-executor/src/executor/initialization.rs:161]: $2080 \text{ bytes} \times 262144 \text{ requests} \approx 545 \text{ mb}$

File: protocol-units/execution/opt-executor/src/executor/services.rs
Comment (l-monninger) [reply protocol-units/execution/opt-executor/src/executor/services.rs:85]: No good reason.

File: networks/suzuka/setup/src/local.rs
Comment (l-monninger) [reply networks/suzuka/setup/src/local.rs:63]: It's been tested iteratively here: https://github.com/movementlabsxyz/infra/pull/126

These updates have been the only time I've hit that 1,000 TPS mark.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs:69]: See my comments above to @mzabaluev. I needed an ID prior to have a `block_id` for the logging. There may be a better way as tracing spans generate their own UID.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs:63]: I thought this would be more likely to actually schedule a green thread. But, reviewing Tokio source, it seems not.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs:266]: Yes, but we were trying things. 

That being said, I haven't seen significant degradation with `spawn_blocking` calls, and after all they just add a closure to the block thread pool. 

For the sake of this PR, so that I don't have to remove all of the internal `spawn_blocking` calls, I can add a `publish_transactions` as I did for `pop_transactions`, so that the overhead of `spawn_blocking` is minimal.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs:111]: No

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs:266]: I did the above. I believe Tokio `spawn_blocking` is particularly good for blocking IO, which `rocksdb` has a lot of. For larger sets and gets like this, I'm an advocate.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (mzabaluev) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs]: Yes, each span already has an internal ID. If you manage to pass the span throughout your flow, you should see it in all events that are descendants in the span (assuming it is enabled), and also in the synthetic close event that I specified the timing subscriber to log. But I don't know if many subscribers display the span ID; also, assigning span IDs is the responsibility of the subscriber and it may in principle not be unique.

I'd favor using explicitly generated IDs if you must have something to identify your flows. But it could be as simple as an (atomic) `u64` counter, no need to invoke a whole crypto-grade RNG for this.

File: protocol-units/execution/opt-executor/src/executor/execution.rs
Comment (mzabaluev) [reply protocol-units/execution/opt-executor/src/executor/execution.rs]: OK, then the code should be removed.

Do we shift the responsibility on the pre-validator, which is also currently conditionally cut out?

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (mzabaluev) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs]: Whichever way it looks clearer to you.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (mzabaluev) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs]: I don't think you can alter the parent span retroactively, but if this event carries `block_id`, it can be correlated with the parent span's `uid`.

File: protocol-units/execution/opt-executor/src/executor/initialization.rs
Comment (mzabaluev) [reply protocol-units/execution/opt-executor/src/executor/initialization.rs:131]: This could be cleaned up in turn, then.

File: protocol-units/execution/opt-executor/src/executor/initialization.rs
Comment (mzabaluev) [inline protocol-units/execution/opt-executor/src/executor/initialization.rs:137]: Another commented out bit.

File: protocol-units/execution/opt-executor/src/executor/initialization.rs
Comment (mzabaluev) [reply protocol-units/execution/opt-executor/src/executor/initialization.rs:161]: OK. I hope we have the memory budget for this in our nodes.

File: protocol-units/execution/opt-executor/src/executor/mod.rs
Comment (mzabaluev) [inline protocol-units/execution/opt-executor/src/executor/mod.rs:31]: This could also be removed.

File: protocol-units/execution/opt-executor/src/executor/mod.rs
Comment (mzabaluev) [inline protocol-units/execution/opt-executor/src/executor/mod.rs:67]: This will no longer work even if uncommented since the let statement for `core_mempool` is gone.

**Review by mzabaluev — COMMENTED** [review]
Looks good from my part; some small cleanup can still be made as commented, and timing logs further cross-correlated and traced with additional spans, as discussed above, but no blockers.
Let @musitdev or anyone else look over the `movement_timing` bits I authored, so we can have the final approval.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (musitdev) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs]: Yes `spawn_blocking` is very important for CPU bounder task, but it's use you be done with care because you can spawn a thousand of them very easily and performance can degrade very fast.
As you don't see perf decrease, it's ok for me. If was mostly to attract attention to this fact.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (musitdev) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs]: I agree with @mzabaluev. A counter can be enough.

File: networks/suzuka/setup/src/local.rs
Comment (musitdev) [reply networks/suzuka/setup/src/local.rs:63]: I was asking about to change the config modification to specific methods. Perhaps you didn't put the comment in the right place.

**Review by musitdev — APPROVED** [review]
As you didn't some performance degradation, it's ok for me.

File: networks/suzuka/setup/src/local.rs
Comment (l-monninger) [reply networks/suzuka/setup/src/local.rs:63]: Yes, that would be a good idea. I think I just responded in the wrong spot. There are quite a few places we can do it though. Should I open up an issue?

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs]: Okay, counter with wrapping add coming up. 

I'll leave the spanning for a future PR.

File: protocol-units/da/m1/light-node/src/v1/sequencer.rs
Comment (l-monninger) [reply protocol-units/da/m1/light-node/src/v1/sequencer.rs]: This was removed, so marking resolved.

File: protocol-units/execution/opt-executor/src/executor/initialization.rs
Comment (l-monninger) [reply protocol-units/execution/opt-executor/src/executor/initialization.rs:161]: We do, but I can tone it down.

File: protocol-units/execution/opt-executor/src/executor/initialization.rs
Comment (l-monninger) [reply protocol-units/execution/opt-executor/src/executor/initialization.rs:131]: Cleaned up.

**Pull request merged**
